<?xml version="1.0"?>
<robot name="basic_arm">

    <!-- Materials -->
    <material name="grey">
        <color rgba='0 0 0 1'/>
    </material>
    
    <material name="green">
        <color rgba="0 0.6 0 1"/>
    </material>
    
    <material name="red">
        <color rgba="0.6 0 0 1"/>
    </material>
    
    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>


    <!-- Base link -->

    <link name="base_link">
        <visual>
            <geometry>
                <box size="0.2 0.3 0.4"/>
            </geometry>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <material name="green"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.2 0.3 0.4"/>
            </geometry>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <mass value="4.0"/>
            <inertia ixx="0.013" ixy="0" ixz="0"
                     iyy="0.01" iyz="0"
                     izz="0.006"/>
        </inertial>
    </link>

    <!-- Base servo -->
    <link name="base_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="-1.5708 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="-1.5708 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- shoulder link -->
    <link name="shoulder_link">
        <visual>
            <geometry>
                <box size="0.1 0.2 0.1"/>
            </geometry>
            <origin xyz="0 0.09 0" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.2 0.1"/>
            </geometry>
            <origin xyz="0 0.09 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0.09 0" rpy="0 0 0"/>
            <mass value="1.5"/>
            <inertia ixx="0.003" ixy="0" ixz="0"
                     iyy="0.002" iyz="0"
                     izz="0.003"/>
        </inertial>
    </link>

    <!-- Shoulder_Servo -->
    <link name="Shoulder_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- elbow link -->
    <link name="elbow_link">
        <visual>
            <geometry>
                <box size="0.1 0.3 0.1"/>
            </geometry>
            <origin xyz="0 0.08 -0.055" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.3 0.1"/>
            </geometry>
            <origin xyz="0 0.08 -0.055" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0.08 -0.055" rpy="0 0 0"/>
            <mass value="1.5"/>
            <inertia ixx="0.004" ixy="0" ixz="0"
                     iyy="0.003" iyz="0"
                     izz="0.004"/>
        </inertial>
    </link>

    <!-- elbow_Servo -->
    <link name="elbow_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- wrist link -->
    <link name="wrist_link">
        <visual>
            <geometry>
                <box size="0.1 0.3 0.04"/>
            </geometry>
            <origin xyz="0 0.16 -0.015" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.3 0.04"/>
            </geometry>
            <origin xyz="0 0.16 -0.015" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0.16 -0.015" rpy="0 0 0"/>
            <mass value="1.0"/>
            <inertia ixx="0.002" ixy="0" ixz="0"
                     iyy="0.001" iyz="0"
                     izz="0.002"/>
        </inertial>
    </link>

    <!-- Right Base Servo -->
    <link name="right_base_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="-1.5708 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="-1.5708 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- right shoulder link -->
    <link name="right_shoulder_link">
        <visual>
            <geometry>
                <box size="0.1 0.2 0.1"/>
            </geometry>
            <origin xyz="0 -0.09 0" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.2 0.1"/>
            </geometry>
            <origin xyz="0 -0.09 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 -0.09 0" rpy="0 0 0"/>
            <mass value="1.5"/>
            <inertia ixx="0.003" ixy="0" ixz="0"
                     iyy="0.002" iyz="0"
                     izz="0.003"/>
        </inertial>
    </link>

    <!-- right_shoulder_servo -->
    <link name="right_shoulder_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- right_elbow_link -->
    <link name="right_elbow_link">
        <visual>
            <geometry>
                <box size="0.1 0.3 0.1"/>
            </geometry>
            <origin xyz="0 -0.08 -0.055" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.3 0.1"/>
            </geometry>
            <origin xyz="0 -0.08 -0.055" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 -0.08 -0.055" rpy="0 0 0"/>
            <mass value="1.5"/>
            <inertia ixx="0.004" ixy="0" ixz="0"
                     iyy="0.003" iyz="0"
                     izz="0.004"/>
        </inertial>
    </link>

    <!-- right_elbow_servo -->
    <link name="right_elbow_servo">
        <visual>
            <geometry>
                <cylinder radius="0.08" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.05"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.3"/>
            <inertia ixx="0.0004" ixy="0" ixz="0"
                     iyy="0.0004" iyz="0"
                     izz="0.0002"/>
        </inertial>
    </link>

    <!-- right_wrist_link -->
    <link name="right_wrist_link">
        <visual>
            <geometry>
                <box size="0.1 0.3 0.04"/>
            </geometry>
            <origin xyz="0 -0.16 -0.015" rpy="0 0 0"/>
            <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.1 0.3 0.04"/>
            </geometry>
            <origin xyz="0 -0.16 -0.015" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 -0.16 -0.015" rpy="0 0 0"/>
            <mass value="1.0"/>
            <inertia ixx="0.002" ixy="0" ixz="0"
                     iyy="0.001" iyz="0"
                     izz="0.002"/>
        </inertial>
    </link>


    <!-- head link -->
    <link name="head_link">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
            <material name="red"/>
        </visual>
        
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
        </collision>
        
        <inertial>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
            <mass value="1.0"/>
            <inertia ixx="0.0008" ixy="0.0" ixz="0.0"
                    iyy="0.0008" iyz="0.0"
                    izz="0.0008"/>
        </inertial>
    </link>


    <!-- left_leg link -->
    <link name="left_leg_link">
        <visual>
            <geometry>
                <box  size="0.1 0.1 0.2"/>
            </geometry>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
            <material name="green"/>
        </visual>
    </link>

    <!-- right_leg link -->
    <link name="right_leg_link">
        <visual>
            <geometry>
                <box  size="0.1 0.1 0.2"/>
            </geometry>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
            <material name="green"/>
        </visual>
    </link>

    <!-- Joints -->

    <joint name="base_to_base_servo_joint" type="continuous">
        <parent link="base_link"/>
        <child link="base_servo"/>
        <origin xyz="0 0.17 0.3" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="base_servo_to_shoulder_joint" type="fixed">
        <parent link="base_servo"/>
        <child link="shoulder_link"/>
        <origin xyz="0 0.03 0" rpy="0 0 0"/>
    </joint>

    <joint name="shoulder_to_shoulder_servo_joint" type="continuous">
        <parent link="shoulder_link"/>
        <child link="Shoulder_servo"/>
        <origin xyz="0 0.15 -0.075" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="shoulder_servo_to_elbow_joint" type="fixed">
        <parent link="Shoulder_servo"/>
        <child link="elbow_link"/>
        <origin xyz="0 0 -0.02" rpy="0 0 0"/>
    </joint>

    <joint name="elbow_to_elbow_servo_joint" type="continuous">
        <parent link="elbow_link"/>
        <child link="elbow_servo"/>
        <origin xyz="0 0.2 -0.13" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="elbow_servo_to_wrist_joint" type="fixed">
        <parent link="elbow_servo"/>
        <child link="wrist_link"/>
        <origin xyz="0 0 -0.03" rpy="0 0 0"/>
    </joint>


    <joint name="base_to_right_base_servo_joint" type="continuous">
        <parent link="base_link"/>
        <child link="right_base_servo"/>
        <origin xyz="0 -0.17 0.3" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="right_base_servo_to_shoulder_joint" type="fixed">
        <parent link="right_base_servo"/>
        <child link="right_shoulder_link"/>
        <origin xyz="0 -0.03 0" rpy="0 0 0"/>
    </joint>  

    <joint name="right_shoulder_to_right_shoulder_servo_joint" type="continuous">
        <parent link="right_shoulder_link"/>
        <child link="right_shoulder_servo"/>
        <origin xyz="0 -0.15 -0.075" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="right_shoulder_servo_to_right_elbow_joint" type="fixed">
        <parent link="right_shoulder_servo"/>
        <child link="right_elbow_link"/>
        <origin xyz="0 0 -0.02" rpy="0 0 0"/>
    </joint>

    <joint name="right_elbow_to_right_elbow_servo_joint" type="continuous">
        <parent link="right_elbow_link"/>
        <child link="right_elbow_servo"/>
        <origin xyz="0 -0.23 -0.13" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="right_elbow_servo_to_right_wrist_joint" type="fixed">
        <parent link="right_elbow_servo"/>
        <child link="right_wrist_link"/>
        <origin xyz="0 0 -0.03" rpy="0 0 0"/>
    </joint>

    <joint name="base_to_head_joint" type="fixed">
       <parent link="base_link"/>
        <child link="head_link"/>
        <origin xyz="0 0 0.4" rpy="0 0 0"/>
    </joint>

    <joint name="base_to_left_leg_joint" type="fixed">
        <parent link="base_link"/>
        <child link="left_leg_link"/>
        <origin xyz="0 -0.1 -0.2" rpy="0 0 0"/>
    </joint>

    <joint name="base_to_right_leg_joint" type="fixed">
        <parent link="base_link"/>
        <child link="right_leg_link"/>
        <origin xyz="0 0.1 -0.2" rpy="0 0 0"/>
    </joint>
</robot>
